src/gc.js

```diff
@@ -256,33 +246,87 @@ function h$gc(t) {
 }
 
 function h$markRetained() {
-    var iter, marked, c, mark = h$gcMark;
+    var iter, marked, w, mark = h$gcMark;
+    var newList = [];
+
+    /*
+      2. Scan the Weak Pointer List. If a weak pointer object has a key that is
+      marked (i.e. reachable), then mark all heap reachable from its value
+      or its finalizer, wand move the weak pointer object to a new list
+    */
     do {
-        TRACE_GC("mark retained iteration");
+        TRACE_GC("mark retained iteration 1/2");
         marked = false;
-        // mark all finalizers of weak references where the key is reachable
-        iter = h$finalizers.iter();
-        while((c = iter.next()) !== null) {
-            if(!IS_MARKED(c.finalizer) && c.m.m === mark) {
-                TRACE_GC("recursively marking weak finalizer: " + h$collectProps(c.finalizer));
-                h$follow(c.finalizer);
-                marked = true;
-            }
-        }
 
-        // mark weak values for reachable keys
-        for(var i=h$scannedWeaks.length-1;i>=0;i--) {
-            var w = h$scannedWeaks[i];
-            if(w.keym.m === mark && w.val !== null && !IS_MARKED(w.val)) {
-                TRACE_GC("marking weak value");
-                h$follow(w.val);
-                marked = true;
-            }
-        }
-        // continue for a next round if we have marked something more
-        // note: this will be slow for very deep chains of weak refs,
-        // change this if that becomes a problem.
+	for (var i = 0; i < h$weakPointerList.length; ++i) {
+	    w = h$weakPointerList[i];
+	    if (w === null) {
+		continue;
+	    }
+	    if (w.keym.m === mark) {
+		TRACE_GC("recursively marking weak: " + h$collectProps(c.finalizer));
+
+		if (w.val !== null && !IS_MARKED(w.val)) {
+		    h$follow(w.val);
+		}
+
+		if (w.finalizer !== null && !IS_MARKED(w.finalizer)) {
+		    h$follow(w.finalizer);
+		}
+
+		newList.push(w);
+		h$weakPointerList[i] = null;
+		marked = true;
+	    }
+	}
+
+        /*
+           3. Repeat from step (2), until a complete scan of Weak Pointer List finds
+              no weak pointer object with a marked keym.
+        */
     } while(marked);
+
+
+    /*
+      4. Scan the Weak Pointer List again. If the weak pointer object is reachable
+         then tombstone it. If the weak pointer object has a finalizer then move
+         it to the Finalization Pending List, and mark all the heap reachable
+         from the finalizer. If the finalizer referes to the key (and/or value),
+         this step will "resurrect" it.
+    */
+    for (var i = 0; i < h$weakPointerList.length; ++i) {
+	w = h$weakPointerList[i];
+	if (w === null) {
+	    continue;
+	}
+
+	TRACE_GC("mark retained iteration 2/2");
+
+	// FIXME: we should set v.val to null ??? aka tombstone
+	//if (w.val !== null && !IS_MARKED(w.val)) {
+	//  w.val = null;
+        //}
+
+	if (w.finalizer !== null) {
+	    if (!IS_MARKED(w.finalizer)) {
+		h$follow(w.finalizer);
```

Comment:
I don't have "If the weak pointer object has a finalizer then move it to the Finalization Pending List", because I assume `h$finalizeWeak` will catch the right finalizers, but I'am unsure about this. Maybe `h$finalizeWeaks` should *only* call those finalizers which have been found here?